keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/AuthorizationSupport.kt

```diff
@@ -18,59 +18,73 @@
+ *
+ * Permission is granted based on the user (as identified by the X-SPINNAKER-USER request header, or X509 client
+ * certificate, or metatron identity) having access to the "service account" associated with keel's entities
+ * (e.g. delivery configs, resources, etc), and that service account having access to the corresponding application
+ * in Spinnaker with the right permission level (READ or WRITE).
+ */
 @Component
 class AuthorizationSupport(
   private val permissionEvaluator: FiatPermissionEvaluator,
   private val repository: KeelRepository
 ) {
   val log: Logger by lazy { LoggerFactory.getLogger(javaClass) }
 
-  fun userCanModifyResource(name: String): Boolean =
-    try {
-      val resource = repository.getResource(name)
-      userCanModifySpec(resource.serviceAccount, resource.id)
-    } catch (e: NoSuchResourceException) {
-      // If resource doesn't exist return true so a 404 is propagated from the controller.
-      true
-    }
-
-  fun userCanModifySpec(serviceAccount: String, specOrName: Any): Boolean {
-    val auth = SecurityContextHolder.getContext().authentication
-    return userCanAccessServiceAccount(auth, serviceAccount, "Resource: $specOrName")
+  enum class Action {
+    READ, WRITE;
+    override fun toString() = name.toLowerCase()
   }
 
-  fun userCanModifyApplication(application: String): Boolean = try {
-    val deliveryConfig = repository.getDeliveryConfigForApplication(application)
-    userCanModifyDeliveryConfig(deliveryConfig.serviceAccount, application, deliveryConfig.name)
-  } catch (e: NoSuchElementException) {
-    // If there are no delivery configs for that application return true so this is handled correctly in the controller.
-    true
+  enum class Entity {
+    APPLICATION, DELIVERY_CONFIG, RESOURCE;
+    override fun toString() = name.toLowerCase()
   }
 
-  fun userCanModifyDeliveryConfig(serviceAccount: String, application: String, manifestName: String): Boolean {
-    val auth = SecurityContextHolder.getContext().authentication
-    return userCanAccessServiceAccount(auth, serviceAccount, "Delivery config for application $application: $manifestName")
-  }
+  data class Permission(val action: Action, val entity: Entity)
 
-  fun userCanAccessServiceAccount(auth: Authentication, serviceAccount: String, specOrName: Any): Boolean {
-    val hasPermission = permissionEvaluator.hasPermission(auth, serviceAccount, "SERVICE_ACCOUNT", "ignored-svcAcct-auth")
-    log.debug(
-      "[AUTH] {} is trying to access service account {}. They{} have permission. {}",
-      auth.principal,
-      serviceAccount,
-      if (hasPermission) "" else " DO NOT",
-      specOrName
-    )
-    return hasPermission
+  fun userCan(action: Action, entity: Entity, identifier: String): Boolean {
+    return try {
+      val auth = SecurityContextHolder.getContext().authentication
+      val serviceAccount = when (entity) {
+        Entity.RESOURCE -> repository.getResource(identifier).serviceAccount
```

Comment:
nit: please add imports for all these so that this is `RESOURCE` instead of `Entity.Resource`